### PR TITLE
introduce SockAddr to store IP addresses

### PR DESCRIPTION
**Pull request by aberaud**

### CODE REVIEW

File: include/opendht/dhtrunner.h [hunk 1]
```diff
@@ -253,7 +254,7 @@ class DhtRunner {
     std::string getStorageLog() const;
     std::string getRoutingTablesLog(sa_family_t af) const;
     std::string getSearchesLog(sa_family_t af = 0) const;
-    std::vector<Address> getPublicAddress(sa_family_t af = 0);
+    std::vector<SockAddr> getPublicAddress(sa_family_t af = 0);
```
Comment (sim590) [inline include/opendht/dhtrunner.h]: I'm noting that this patch does modify the API.

File: include/opendht/network_engine.h [hunk 2]
```diff
@@ -196,7 +196,7 @@ class NetworkEngine final {
      * @param saddr (type: sockaddr*) sockaddr* pointer containing address ip information.
```
Comment (sim590) [inline include/opendht/network_engine.h]: Missing `#include "sockaddr.h"` ?

File: include/opendht/dht.h [hunk 3]
```diff
@@ -113,14 +113,20 @@ class Dht {
      * The node is not pinged, so this should be
```
Comment (sim590) [inline include/opendht/dht.h]: Missing `#include "sockaddr.h"` ?

File: include/opendht/node_cache.h [hunk 4]
```diff
@@ -28,7 +28,7 @@ namespace dht {
```
Comment (sim590) [inline include/opendht/node_cache.h]: Missing `#include "sockaddr.h"` ?

File: src/network_engine.cpp [hunk 5]
```diff
@@ -253,37 +253,25 @@ NetworkEngine::blacklistNode(const std::shared_ptr<Node>& n)
...
-    /*if (isBlacklisted(sa, salen))
-        return true;*/
-
-    for (unsigned i = 0; i < BLACKLISTED_MAX; i++) {
-        if (memcmp(&blacklist[i], sa, salen) == 0)
-            return true;
-    }
-
-    return false;
+    return blacklist.find(addr) != blacklist.end();
```
Comment (sim590) [inline src/network_engine.cpp]: Cool. This is shorter :wink:.

File: src/network_engine.cpp [hunk 6]
```diff
@@ -652,41 +640,44 @@ NetworkEngine::deserializeNodes(ParsedMessage& msg) {
...
-            sin.sin_family = AF_INET;
-            memcpy(&sin.sin_addr, ni + ni_id.size(), 4);
-            memcpy(&sin.sin_port, ni + ni_id.size() + 4, 2);
-            if (isMartian((sockaddr*)&sin, sizeof(sin)) || isNodeBlacklisted((sockaddr*)&sin, sizeof(sin)))
+            auto sin = (sockaddr_in*)&addr.first;
+            std::fill_n((uint8_t*)sin, sizeof(sockaddr_in), 0);
+            sin->sin_family = AF_INET;
+            memcpy(&sin->sin_addr, ni + ni_id.size(), 4);
+            memcpy(&sin->sin_port, ni + ni_id.size() + 4, 2);
+            addr.second = sizeof(sockaddr_in);
```
Comment (sim590) [inline src/network_engine.cpp]: Could this be encapsulated into one of `SockAddr`'s constructor in some way?

File: src/network_engine.cpp [hunk 7]
```diff
@@ -652,41 +640,44 @@ NetworkEngine::deserializeNodes(ParsedMessage& msg) {
...
-            sin6.sin6_family = AF_INET6;
-            memcpy(&sin6.sin6_addr, ni + HASH_LEN, 16);
-            memcpy(&sin6.sin6_port, ni + HASH_LEN + 16, 2);
-            if (isMartian((sockaddr*)&sin6, sizeof(sin6)) || isNodeBlacklisted((sockaddr*)&sin6, sizeof(sin6)))
+            auto sin6 = (sockaddr_in6*)&addr.first;
+            std::fill_n((uint8_t*)sin6, sizeof(sockaddr_in6), 0);
+            sin6->sin6_family = AF_INET6;
+            memcpy(&sin6->sin6_addr, ni + HASH_LEN, 16);
+            memcpy(&sin6->sin6_port, ni + HASH_LEN + 16, 2);
+            addr.second = sizeof(sockaddr_in6);
```
Comment (sim590) [inline src/network_engine.cpp]: ^

File: include/opendht/dhtrunner.h (hunk 1 above)
Comment (aberaud) [inline include/opendht/dhtrunner.h]: yes

**Pull request merged**